kauri/kauri.go

```diff
@@ -0,0 +1,320 @@
+				View:      uint64(k.currentView),
+			})
+		}
+	}
+}
+
+type serviceImpl struct {
+	k *Kauri
+}
+
+func (i serviceImpl) SendContribution(ctx gorums.ServerCtx, request *kauripb.Contribution) {
+	i.k.eventLoop.AddEvent(ContributionRecvEvent{Contribution: request})
+}
+
+// ContributionRecvEvent is raised when a contribution is received.
+type ContributionRecvEvent struct {
+	Contribution *kauripb.Contribution
+}
+
+func (k *Kauri) canMergeContributions(a, b hotstuff.QuorumSignature) bool {
+	canMerge := true
+	if a == nil || b == nil {
+		k.logger.Info("one of it is nil")
+		return false
+	}
+	a.Participants().RangeWhile(func(i hotstuff.ID) bool {
+		b.Participants().RangeWhile(func(j hotstuff.ID) bool {
+			// cannot merge a and b if they both contain a contribution from the same ID.
+			if i == j {
+				canMerge = false
+			}
+			return canMerge
+		})
+		return canMerge
+	})
+	return canMerge
+}
+
+func (k *Kauri) verifyContribution(signature hotstuff.QuorumSignature, hash hotstuff.Hash) bool {
+	verified := false
+	block, ok := k.blockChain.Get(hash)
+	if !ok {
+		k.logger.Info("failed to fetch the block ", hash)
+		return verified
+	}
+	verified = k.crypto.Verify(signature, block.ToBytes())
+	return verified
+}
+
+func (k *Kauri) mergeWithContribution(currentSignature hotstuff.QuorumSignature) (bool, error) {
+	isVerified := k.verifyContribution(currentSignature, k.blockHash)
+	if !isVerified {
+		k.logger.Info("Contribution verification failed for view ", k.currentView,
+			"from participants", currentSignature.Participants(), " block hash ", k.blockHash)
+		return false, errors.New("unable to verify the contribution")
+	}
+	if k.aggregatedContribution == nil {
+		k.aggregatedContribution = currentSignature
+		return false, nil
+	}
+
+	if k.canMergeContributions(currentSignature, k.aggregatedContribution) {
+		new, err := k.crypto.Combine(currentSignature, k.aggregatedContribution)
+		if err == nil {
+			k.aggregatedContribution = new
+			if new.Participants().Len() >= k.configuration.QuorumSize() {
+				k.logger.Debug("Aggregated Complete QC and sending the event")
+				k.eventLoop.AddEvent(hotstuff.NewViewMsg{
+					SyncInfo: hotstuff.NewSyncInfo().WithQC(hotstuff.NewQuorumCert(
+						k.aggregatedContribution,
+						k.currentView,
+						k.blockHash,
+					)),
+				})
+				return true, nil
+			}
+		} else {
+			k.logger.Info("Failed to combine signatures: %v", err)
+			return false, errors.New("unable to combine signature")
+		}
+	} else {
+		k.logger.Debug("Failed to merge signatures due to overlap of signatures.")
+		return false, errors.New("unable to merge signature")
+	}
+	return false, nil
+}
+
+func isSubSet(a, b []hotstuff.ID) bool {
+	c := hotstuff.NewIDSet()
+	for _, id := range b {
+		c.Add(id)
+	}
+	for _, id := range a {
+		if !c.Contains(id) {
+			return false
+		}
+	}
+	return true
+}
+
+func (k *Kauri) randomizeIDS(hash hotstuff.Hash, leaderID hotstuff.ID) map[hotstuff.ID]int {
```

Comment:
Suggest to rename this to `shuffleTreeNodes` and move it into the tree. And instead of passing in the `hash` you could move the `seed` calculation out from this method... so that the signature becomes:
```go
func (t *faultFreeTree) shuffleTreeNodes(seed int64, leaderID hotstuff.ID)
```
This way you can update the internals of the tree directly avoiding the passing around of maps from this method to `InitializeWithPIDs` (which could be removed; its functionality can be done in the constructor and in the above method.)